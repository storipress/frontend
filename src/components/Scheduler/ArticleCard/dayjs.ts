export { dayjs } from '~/lib/dayjs'
