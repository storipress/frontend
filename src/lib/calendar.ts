import type { dayjs } from './dayjs'

export interface CalendarDate {
  date: string
  day: dayjs.Dayjs
  isCurrentWeek?: boolean
  isCurrentMonth?: boolean
  isNextMonth?: boolean
  isToday?: boolean
}

export function getWholeMonth(currentDay: dayjs.Dayjs, today: dayjs.Dayjs): CalendarDate[] {
  const daysInMonth = currentDay.daysInMonth()
  const firstDay = currentDay.date(1)
  const lastDay = currentDay.date(daysInMonth)

  const firstDayOfWeek = firstDay.day()
  const lastDayOfWeek = lastDay.day()

  const days: dayjs.Dayjs[] = []

  if (firstDayOfWeek === 0) {
    for (let i = 6; i > 0; i--) {
      days.push(firstDay.subtract(i, 'day'))
    }
  } else {
    // we are start at Monday
    for (let i = 1; i < firstDayOfWeek; i++) {
      days.push(firstDay.subtract(firstDayOfWeek - i, 'day'))
    }
  }

  for (let i = 0; i < daysInMonth; i++) {
    days.push(firstDay.add(i, 'day'))
  }

  for (let i = 0; i < 7 - lastDayOfWeek; i++) {
    days.push(lastDay.add(i + 1, 'day'))
  }

  return days.map((day) => ({
    day,
    date: day.format('YYYY-MM-DD'),
    isCurrentWeek: day.isSame(today, 'week'),
    isCurrentMonth: day.month() === currentDay.month(),
    // Only show the month name if it's the first day of the next month
    isNextMonth: day.month() === currentDay.month() + 1 && day.date() === 1,
    isToday: day.isSame(today, 'day'),
  }))
}
