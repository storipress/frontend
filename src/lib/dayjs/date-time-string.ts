import type { PluginFunc } from 'dayjs'

declare module 'dayjs' {
  function fromDateTimeString(dateTime: string): Dayjs

  interface Dayjs {
    toDateTimeString: () => string
  }
}

const SLICE_LENGTH = -5 // -'.sssZ'.length

export const dateTimeString: PluginFunc = (_, Dayjs, factory) => {
  factory.fromDateTimeString = (dateTime) => {
    return factory(dateTime)
  }

  Dayjs.prototype.toDateTimeString = function (): string {
    return `${this.toISOString().slice(0, SLICE_LENGTH)}Z`
  }
}
