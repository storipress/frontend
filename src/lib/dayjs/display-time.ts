import type { PluginFunc } from 'dayjs'

declare module 'dayjs' {
  interface Dayjs {
    toDisplayString: () => string
  }
}

export const toDisplayString: PluginFunc = (_, Dayjs, dayjs) => {
  Dayjs.prototype.toDisplayString = function (): string {
    const today = dayjs().startOf('day')
    const yesterday = today.subtract(1, 'day')
    const lastWeek = today.subtract(7, 'day')
    if (this.isBetween(today, today.endOf('day'))) {
      return this.format('kk:mm [Today]')
    }
    if (this.isBetween(yesterday, yesterday.endOf('day'))) {
      return this.format('kk:mm [Yesterday]')
    }
    if (this.isBetween(lastWeek, today)) {
      return this.format('kk:mm ddd')
    }

    return this.format('kk:mm ddd MMM')
  }
}
