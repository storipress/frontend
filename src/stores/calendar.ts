import { defineStore } from 'pinia'

export const useCalendarStore = defineStore('calendar', {
  state: () => ({
    currentIndex: undefined as number | undefined,
  }),
  actions: {
    setIndex(index: number) {
      this.currentIndex = index
    },
    resetIndex() {
      this.currentIndex = undefined
    },
  },
})
