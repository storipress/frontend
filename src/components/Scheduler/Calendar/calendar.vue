<script lang="ts" setup>
import { dayjs } from '../ArticleCard/dayjs'
import type { Article, ScheduledArticle, ScheduledArticleWithDay } from '../definitions'
import { ViewTypes } from '../definitions'
import { useSchedulerStore } from '../store'
import CalendarHeader from './calendar-header.vue'
import { isScheduledArticle } from './helpers'
import MonthBody from './month-body.vue'
import WeekBody from './week-body.vue'

const props = withDefaults(
  defineProps<{
    today?: dayjs.Dayjs
    articles?: Article[]
    viewType?: ViewTypes
    scrollToNow?: boolean
    newArticle?: Article
    isLoading?: boolean
  }>(),
  {
    today: () => dayjs(),
    articles: () => [],
    viewType: ViewTypes.Month,
  },
)

const emit = defineEmits<{
  (event: 'update:viewType', viewType: string): void
  (event: 'update:newArticle', newArticle?: Article | undefined): void
}>()

const modelNewArticle = useVModel(props, 'newArticle', emit)

const store = useSchedulerStore()

watch(
  () => props.today,
  (today) => {
    store.currentDay = today
  },
  { immediate: true },
)

const modelViewType = useVModel(props, 'viewType', emit, { passive: true })

const scheduledArticles = computed<ScheduledArticleWithDay[]>(() =>
  props.articles
    .filter((article): article is ScheduledArticle => isScheduledArticle(article))
    .map(
      (article): ScheduledArticleWithDay => ({
        ...article,
        day: store.parseDate(article.scheduledAt),
      }),
    ),
)

// bind to store, so we can control it at another place
const currentDay = toRef(store, 'currentDay')

function addMonth(amount: number) {
  currentDay.value = currentDay.value.add(amount, 'month')
}

function addWeek(amount: number) {
  currentDay.value = currentDay.value.add(amount * 5, 'day')
}

function handlePage(amount: number) {
  if (modelViewType.value === ViewTypes.Month) {
    addMonth(amount)
  } else {
    addWeek(amount)
  }
}

function handleWeekView(day: dayjs.Dayjs) {
  currentDay.value = day
  modelViewType.value = ViewTypes.FiveDay
}
</script>

<template>
  <div class="layer-2 relative flex max-h-full flex-col bg-white">
    <CalendarHeader
      v-model:view-type="modelViewType"
      :day="currentDay"
      @click-previous="handlePage(-1)"
      @click-today="currentDay = today"
      @click-next="handlePage(1)"
    />

    <MonthBody
      v-if="modelViewType === ViewTypes.Month"
      v-model:new-article="modelNewArticle"
      :today="today"
      :current-day="currentDay"
      :articles="scheduledArticles"
      :is-loading="isLoading"
      @week-view="handleWeekView"
    />
    <WeekBody
      v-else
      v-model:new-article="modelNewArticle"
      :current-day="currentDay"
      :today="today"
      :articles="scheduledArticles"
      :scroll-to-now="scrollToNow"
    />
  </div>
</template>
