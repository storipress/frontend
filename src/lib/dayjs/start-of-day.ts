import dayjs from 'dayjs'
import type { Dayjs, PluginFunc } from 'dayjs'

declare module 'dayjs' {
  interface Dayjs {
    startOfDay: () => Dayjs
    $x: {
      $timezone: string
    }
  }
}

// workaround of UTC bug
// ref: test/dayjs.test.ts
export const startOfDay: PluginFunc = (_, DayjsClass) => {
  DayjsClass.prototype.startOfDay = function (): Dayjs {
    return dayjs.tz(this.startOf('day').format('YYYY-MM-DD HH:mm:ss:SSS'), this.$x.$timezone)
  }
}
