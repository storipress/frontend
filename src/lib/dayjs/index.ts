import dayjs from 'dayjs'
import customParseFormat from 'dayjs/plugin/customParseFormat'
import utc from 'dayjs/plugin/utc'
import timezone from 'dayjs/plugin/timezone'
import isBetween from 'dayjs/plugin/isBetween'
import advancedFormat from 'dayjs/plugin/advancedFormat'
import updateLocale from 'dayjs/plugin/updateLocale'
import { dateTimeString } from './date-time-string'
import { toDisplayString } from './display-time'
import { startOfDay } from './start-of-day'

dayjs.extend(customParseFormat)
dayjs.extend(utc)
dayjs.extend(advancedFormat)
dayjs.extend(isBetween)
dayjs.extend(timezone)
dayjs.extend(dateTimeString)
dayjs.extend(toDisplayString)
dayjs.extend(startOfDay)
dayjs.extend(updateLocale)

dayjs.updateLocale('en', {
  weekStart: 1,
})

export { dayjs }
