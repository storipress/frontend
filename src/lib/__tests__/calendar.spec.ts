import { expect, it } from 'vitest'
import { dedent } from 'ts-dedent'
import { dayjs } from '../dayjs'
import { getWholeMonth } from '../calendar'

// test helper function
function convertCalendarToDays(day: dayjs.Dayjs, cal: string) {
  let foundFirstDay = 0
  const previous = day.subtract(1, 'month').date(1)
  const next = day.add(1, 'month').date(1)
  const dates = cal.replaceAll(/[\n ]+/g, ' ').split(' ')
  const days = []

  for (const date of dates) {
    const d = Number.parseInt(date, 10)
    if (d === 1) {
      foundFirstDay += 1
    }
    if (foundFirstDay === 0) {
      // previous month
      days.push(previous.set('date', d))
    } else if (foundFirstDay === 1) {
      // current month
      days.push(day.set('date', d))
    } else if (foundFirstDay === 2) {
      // next month
      days.push(next.set('date', d))
    }
  }

  return days
}

// test case is generate by unix cal utility `ncal -1Mbm <month>`
describe('convertCalendarToDays', () => {
  it('convertCalendarToDays function work', () => {
    const date = dayjs('2022-01-01')
    const cal = dedent`
  27 28 29 30 31  1  2
  3  4  5  6  7  8  9
  10 11 12 13 14 15 16
  17 18 19 20 21 22 23
  24 25 26 27 28 29 30
  31 1  2  3  4  5  6
  `

    const days = convertCalendarToDays(date, cal)
    expect(days).toHaveLength(42)
    expect(days.map((d) => d.format('YYYY-MM-DD'))).toMatchSnapshot()
  })

  it('convertCalendarToDays function work if start at 1', () => {
    const date = dayjs('2022-08-01')
    const cal = dedent`
  1  2  3  4  5  6  7
  8  9 10 11 12 13 14
  15 16 17 18 19 20 21
  22 23 24 25 26 27 28
  29 30 31  1  2  3  4
  `

    const days = convertCalendarToDays(date, cal)
    expect(days).toHaveLength(35)
    expect(days.map((d) => d.format('YYYY-MM-DD'))).toMatchSnapshot()
  })
})

it.each([
  [
    '2022-01-01',
    dedent`
  27 28 29 30 31  1  2
  3  4  5  6  7  8  9
  10 11 12 13 14 15 16
  17 18 19 20 21 22 23
  24 25 26 27 28 29 30
  31 1  2  3  4  5  6
  `,
  ],
  [
    '2022-06-01',
    dedent`
30  31  1  2  3  4  5
 6  7  8  9 10 11 12
13 14 15 16 17 18 19
20 21 22 23 24 25 26
27 28 29 30  1  2  3`,
  ],
  [
    '2022-08-01',
    dedent`
  1  2  3  4  5  6  7
  8  9 10 11 12 13 14
  15 16 17 18 19 20 21
  22 23 24 25 26 27 28
  29 30 31  1  2  3  4
  `,
  ],
])('getWholeMonth work correctly with %s', (date, cal) => {
  const day = dayjs(date)
  const expectDays = convertCalendarToDays(day, cal)

  const month = getWholeMonth(day, day)
  expect(month).toHaveLength(expectDays.length)
  // check calendar days are correct
  for (const [i, { day }] of month.entries()) {
    expect(day.isSame(expectDays[i], 'day')).toBe(true)
  }
})
