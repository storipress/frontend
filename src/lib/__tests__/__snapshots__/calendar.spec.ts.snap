// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertCalendarToDays > convertCalendarToDays function work 1`] = `
[
  "2021-12-27",
  "2021-12-28",
  "2021-12-29",
  "2021-12-30",
  "2021-12-31",
  "2022-01-01",
  "2022-01-02",
  "2022-01-03",
  "2022-01-04",
  "2022-01-05",
  "2022-01-06",
  "2022-01-07",
  "2022-01-08",
  "2022-01-09",
  "2022-01-10",
  "2022-01-11",
  "2022-01-12",
  "2022-01-13",
  "2022-01-14",
  "2022-01-15",
  "2022-01-16",
  "2022-01-17",
  "2022-01-18",
  "2022-01-19",
  "2022-01-20",
  "2022-01-21",
  "2022-01-22",
  "2022-01-23",
  "2022-01-24",
  "2022-01-25",
  "2022-01-26",
  "2022-01-27",
  "2022-01-28",
  "2022-01-29",
  "2022-01-30",
  "2022-01-31",
  "2022-02-01",
  "2022-02-02",
  "2022-02-03",
  "2022-02-04",
  "2022-02-05",
  "2022-02-06",
]
`;

exports[`convertCalendarToDays > convertCalendarToDays function work if start at 1 1`] = `
[
  "2022-08-01",
  "2022-08-02",
  "2022-08-03",
  "2022-08-04",
  "2022-08-05",
  "2022-08-06",
  "2022-08-07",
  "2022-08-08",
  "2022-08-09",
  "2022-08-10",
  "2022-08-11",
  "2022-08-12",
  "2022-08-13",
  "2022-08-14",
  "2022-08-15",
  "2022-08-16",
  "2022-08-17",
  "2022-08-18",
  "2022-08-19",
  "2022-08-20",
  "2022-08-21",
  "2022-08-22",
  "2022-08-23",
  "2022-08-24",
  "2022-08-25",
  "2022-08-26",
  "2022-08-27",
  "2022-08-28",
  "2022-08-29",
  "2022-08-30",
  "2022-08-31",
  "2022-09-01",
  "2022-09-02",
  "2022-09-03",
  "2022-09-04",
]
`;
